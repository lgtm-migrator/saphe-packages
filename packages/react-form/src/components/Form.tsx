import React, {
  ChangeEvent,
  Dispatch,
  FormEvent,
  ReactElement,
  SetStateAction,
  useState,
} from 'react';
import { Config as HookConfig } from '../hooks/useForm';
import useRecaptcha from '../hooks/useRecaptcha';
import { FieldTypes } from '../utils/fieldTypes';
import {
  FormState,
  getInitialFormState,
  formatFieldValue,
  getDefaultFieldValue,
} from '../utils/formHelpers';
import { Fields, FormValues, HTMLField } from '../utils/helperTypes';
import { validateField } from '../utils/validationHelpers';
import { ValidationModes } from '../utils/validationTypes';
import Field from './Field';
import FormFieldContainer from './helpers/FormFieldContainer';
import SubmitButton from './helpers/SubmitButton';

interface Props<T extends Fields> extends HookConfig<T> {
  isSubmitting: boolean;
  setIsSubmitting: Dispatch<SetStateAction<boolean>>;
}

export default function Form<T extends Fields>(props: Props<T>): ReactElement {
  const {
    name,
    fieldPack,
    validationMode = ValidationModes.AFTER_BLUR,
    fields,
    recaptcha,
    onSubmit,
    isSubmitting,
    setIsSubmitting,
  } = props;

  const [formState, setFormState] = useState<FormState>(
    getInitialFormState(fields),
  );
  const { Recaptcha, recaptchaToken } = useRecaptcha(recaptcha);

  const handleSubmit = async (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    let canSubmit = true;
    const errors: Record<string, string> = {};
    const touched: Record<string, boolean> = {};
    for (const [fieldName, field] of Object.entries(fields)) {
      const error = validateField(field, formState.values[fieldName]);
      errors[fieldName] = error;
      touched[fieldName] = true;
      if (error !== '') canSubmit = false;
    }

    if (!canSubmit) setFormState({ ...formState, errors, touched });
    else if (!!recaptcha && !recaptchaToken)
      console.error(recaptcha.errorMessage);
    else if (onSubmit) {
      setIsSubmitting(true);

      const submitValues: Record<string, string | number | boolean> = {};
      for (const [fieldName, field] of Object.entries(fields))
        submitValues[fieldName] = formatFieldValue(
          field,
          formState.values[fieldName] ?? '',
        );

      Promise.resolve(
        onSubmit(submitValues as FormValues<T>, { recaptchaToken }),
      );

      setIsSubmitting(false);
    }
  };

  const handleChange = (e: ChangeEvent<HTMLField>, fieldName: string) => {
    const targetValue =
      fields[fieldName]?.type === FieldTypes.CHECKBOX
        ? String((e as ChangeEvent<HTMLInputElement>).target.checked)
        : e.target.value;
    let error = formState.errors[fieldName] ?? '';

    if (
      validationMode === ValidationModes.ON_CHANGE ||
      (validationMode === ValidationModes.AFTER_BLUR &&
        formState.touched[fieldName])
    )
      error = validateField(fields[fieldName], targetValue);

    setFormState({
      ...formState,
      errors: {
        ...formState.errors,
        [fieldName]: error,
      },
      values: {
        ...formState.values,
        [fieldName]: targetValue,
      },
    });
  };

  const handleBlur = (fieldName: string) => {
    let error = formState.errors[fieldName] ?? '';
    if (
      validationMode === ValidationModes.ON_BLUR ||
      validationMode === ValidationModes.AFTER_BLUR
    )
      error = validateField(fields[fieldName], formState.values[fieldName]);

    setFormState({
      ...formState,
      errors: {
        ...formState.errors,
        [fieldName]: error,
      },
      touched: {
        ...formState.touched,
        [fieldName]: true,
      },
    });
  };

  return (
    <form onSubmit={(e) => handleSubmit(e)}>
      {Object.keys(fields).map((fieldName, index) => {
        const field = fields[fieldName];
        const fieldId = `${name}${fieldName
          .charAt(0)
          .toUpperCase()}${fieldName.slice(1)}`;
        if (field === undefined) return <div />;
        return (
          <Field
            key={index}
            id={fieldId}
            describedBy={`${fieldId}Description`}
            field={field}
            fieldPack={fieldPack}
            name={fieldName}
            description={field.description}
            error={formState.errors[fieldName] ?? ''}
            value={
              formState.values[fieldName]?.toString() ??
              getDefaultFieldValue(field).toString()
            }
            onChange={(e) => handleChange(e, fieldName)}
            onBlur={() => handleBlur(fieldName)}
          />
        );
      })}

      {recaptcha && (
        <FormFieldContainer fieldPack={fieldPack}>
          <Recaptcha />
        </FormFieldContainer>
      )}

      <SubmitButton {...{ isSubmitting, fieldPack }} />
    </form>
  );
}
