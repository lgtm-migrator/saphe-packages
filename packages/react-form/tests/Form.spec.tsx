import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import Form from '../src/components/Form';
import {
  minimalNumberField,
  minimalCheckField,
  minimalSelectField,
  minimalTextAreaField,
  minimalTextField,
  matchSnapshot,
} from './testHelpers';

const fields = {
  text: minimalTextField,
  textArea: minimalTextAreaField,
  select: minimalSelectField,
  check: minimalCheckField,
  number: minimalNumberField,
};

describe('Form', () => {
  it.todo('tests');
  //   it('renders', () => {
  //     matchSnapshot(
  //       <Form isSubmitting={false} setIsSubmitting={jest.fn()} name="testForm" fields={fields} />,
  //     );
  //   });

  //   it('submits', async () => {
  //     const setIsSubmittingMock = jest.fn();
  //     const onSubmitMock = jest.fn();

  //     render(
  //       <Form
  //         isSubmitting={false}
  //         setIsSubmitting={setIsSubmittingMock}
  //         onSubmit={onSubmitMock}
  //         name="testForm"
  //         fields={fields}
  //       />,
  //     );

  //     userEvent.click(screen.getByText('Submit'));
  //     expect(onSubmitMock.mock.calls.length).toBe(1);
  //     expect(setIsSubmittingMock.mock.calls.length).toBe(2);
  //   });

  //   it('supports change events', () => {
  //     render(
  //       <Form
  //         isSubmitting={false}
  //         setIsSubmitting={jest.fn()}
  //         onSubmit={jest.fn()}
  //         name="testForm"
  //         fields={fields}
  //       />,
  //     );

  //     const textField = screen.getByLabelText('Text Field') as HTMLInputElement;
  //     userEvent.click(textField);
  //     userEvent.type(textField, 'Text input');
  //     expect(textField.value).toBe('Text input');

  //     const textAreaField = screen.getByLabelText('Text Area Field') as HTMLTextAreaElement;
  //     userEvent.click(textAreaField);
  //     userEvent.type(textAreaField, 'Text area input');
  //     expect(textAreaField.value).toBe('Text area input');

  //     const selectField = screen.getByLabelText('Select Field') as HTMLTextAreaElement;
  //     expect(selectField.value).toBe('');
  //     userEvent.selectOptions(selectField, 'option2');
  //     expect(selectField.value).toBe('option2');

  //     const numberField = screen.getByLabelText('Number Field') as HTMLInputElement;
  //     userEvent.type(numberField, '1234d');
  //     expect(numberField.value).toBe('1234');

  //     const checkField = screen.getByLabelText('Check Field') as HTMLInputElement;
  //     expect(checkField.checked).toBeFalsy();
  //     userEvent.click(checkField);
  //     expect(checkField.checked).toBeTruthy();
  //   });

  //   it('validates required fields', () => {
  //     const requiredText = 'This field is required';
  //     const required = {
  //       validation: {
  //         required: requiredText,
  //       },
  //     };

  //     render(
  //       <Form
  //         isSubmitting={false}
  //         setIsSubmitting={jest.fn()}
  //         onSubmit={jest.fn()}
  //         name="testForm"
  //         fields={{
  //           text: { ...minimalTextField, ...required },
  //           textArea: { ...minimalTextAreaField, ...required },
  //           select: { ...minimalSelectField, ...required },
  //           check: { ...minimalCheckField, ...required },
  //           number: { ...minimalNumberField, ...required },
  //         }}
  //       />,
  //     );

  //     const textErrorTestId = 'testFormTextError';
  //     const textField = screen.getByLabelText('Text Field');
  //     expect(screen.queryByTestId(textErrorTestId)).toBeNull();
  //     fireEvent.blur(textField);
  //     expect(screen.getByTestId(textErrorTestId).innerHTML).toBe(requiredText);

  //     // TODO other fields
  //   });

  //   it.todo('validates string');
});
