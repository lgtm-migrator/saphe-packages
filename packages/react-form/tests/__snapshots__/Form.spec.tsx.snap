// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "marginBottom": "25px",
      }
    }
  >
    <label
      htmlFor="testFormText"
    >
      Text Field
    </label>
    <input
      aria-describedby="testFormTextDescription"
      disabled={false}
      id="testFormText"
      name="text"
      onBlur={[Function]}
      onChange={[Function]}
      style={
        Object {
          "borderColor": "#767676",
          "borderRadius": "3px",
          "borderStyle": "solid",
          "borderWidth": "1px",
          "marginBottom": "3px",
        }
      }
      type="text"
      value=""
    />
    
    
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "marginBottom": "25px",
      }
    }
  >
    <label
      htmlFor="testFormTextArea"
    >
      Text Area Field
    </label>
    <textarea
      aria-describedby="testFormTextAreaDescription"
      disabled={false}
      id="testFormTextArea"
      name="textArea"
      onBlur={[Function]}
      onChange={[Function]}
      rows={6}
      style={
        Object {
          "borderColor": "#767676",
          "borderRadius": "3px",
          "borderStyle": "solid",
          "borderWidth": "1px",
          "marginBottom": "3px",
        }
      }
      value=""
    />
    
    
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "marginBottom": "25px",
      }
    }
  >
    <label
      htmlFor="testFormSelect"
    >
      Select Field
    </label>
    <select
      aria-describedby="testFormSelectDescription"
      disabled={false}
      id="testFormSelect"
      name="select"
      onBlur={[Function]}
      onChange={[Function]}
      style={
        Object {
          "borderColor": "#767676",
          "borderRadius": "3px",
          "borderStyle": "solid",
          "borderWidth": "1px",
          "marginBottom": "3px",
        }
      }
      value=""
    >
      <option
        disabled={true}
        value=""
      >
        
      </option>
      <option
        value="option1"
      >
        Option 1
      </option>
      <option
        value="option2"
      >
        Option 2
      </option>
    </select>
    
    
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "marginBottom": "25px",
      }
    }
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
        }
      }
    >
      <input
        aria-describedby="testFormCheckDescription"
        disabled={false}
        id="testFormCheck"
        name="check"
        onBlur={[Function]}
        onChange={[Function]}
        style={
          Object {
            "borderColor": "#767676",
            "borderRadius": "3px",
            "borderStyle": "solid",
            "borderWidth": "1px",
            "marginBottom": "3px",
            "marginRight": "10px",
          }
        }
        type="checkbox"
        value="false"
      />
      <label
        htmlFor="testFormCheck"
      >
        Check Field
      </label>
    </div>
    
    
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "marginBottom": "25px",
      }
    }
  >
    <label
      htmlFor="testFormNumber"
    >
      Number Field
    </label>
    <input
      aria-describedby="testFormNumberDescription"
      disabled={false}
      id="testFormNumber"
      name="number"
      onBlur={[Function]}
      onChange={[Function]}
      style={
        Object {
          "borderColor": "#767676",
          "borderRadius": "3px",
          "borderStyle": "solid",
          "borderWidth": "1px",
          "marginBottom": "3px",
        }
      }
      type="number"
      value=""
    />
    
    
  </div>
  <button
    disabled={false}
    type="submit"
  >
    Submit
  </button>
</form>
`;
